add FirstCallColumn that prints the execution time of the first call

adamsitnik opened the pull request:
Sample:

```cs
class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args,
        DefaultConfig.Instance.With(new FirstCallColumn()));
}

public class RealColdStart
{
    private bool firstCall;

    [Benchmark]
    public void Foo()
    {
        if (firstCall == false)
        {
            firstCall = true;
            Console.WriteLine("// First call");
            Thread.Sleep(1000);
        }
        else
            Thread.Sleep(10);
    }
}
```

BenchmarkDotNet=v0.12.0.20191204-develop, OS=Windows 10.0.18363
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
  [Host]     : .NET Framework 4.8 (4.8.4018.0), X64 RyuJIT
  DefaultJob : .NET Framework 4.8 (4.8.4018.0), X64 RyuJIT


| Method |     Mean |    Error |   StdDev |   Median | First Call |
|------- |---------:|---------:|---------:|---------:|-----------:|
|    Foo | 14.31 ms | 0.561 ms | 1.498 ms | 14.88 ms |   1.0142 s |

cc @eiriktsarpalis

AndreyAkinshin commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public bool AlwaysShow => true;
+
+        public ColumnCategory Category => ColumnCategory.Custom;
```
I think it will be better to use the `Statistics` category with the "latest" priority. From my point of view, this column may be considered as one of the statistics columns, it shouldn't be placed after baseline columns.

AndreyAkinshin commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,48 @@
+            if (!summary.HasReport(benchmarkCase) || !summary[benchmarkCase].Success || !summary[benchmarkCase].AllMeasurements.Any(IsFirstCall))
+                return "-";
+
+            var measurement = summary[benchmarkCase].AllMeasurements.Single(IsFirstCall);
```
Theoretically, the Engine may perform several JittingWorkload iterations that will lead to an exception during summary table building. It will be better to select all `IsFirstCall` measurements, sort them by IterationStage, then by `LaunchIndex`, and take the first measurement. If you think that a situation with several "FirstCall" measurements is an error that should be reported, please use analyzers instead; don't throw exceptions in `IColumn.GetValue`.

adamsitnik replied:
good point, I've added sorting

AndreyAkinshin commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,48 @@
+﻿using System;
```
The `System`, `System.Collections.Generic`, and `System.Text` namespaces are unused, they can be safily removed.

adamsitnik commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public bool IsDefault(Summary summary, BenchmarkCase benchmarkCase) => false;
+
+        private static bool IsFirstCall(Measurement m)
```
Great catch, fixed!

adamsitnik commented:
@AndreyAkinshin I've fixed all the issues and added docs. 

I did not add an analyzer because there is only one situation when we don't have the first call result: the benchmark has failed to run (obvious what `-` means).

AndreyAkinshin commented on docs/articles/samples/IntroFirstCallColumn.md:
````diff
@@ -0,0 +1,58 @@
+```cs
+// attribute
+[FirstCallColumn]
+public class MyBenchmarkClass
````
The benchmark body is missed.

adamsitnik replied:
@AndreyAkinshin the body will be included below by `[!code-csharp[IntroColdStart.cs](../../../samples/BenchmarkDotNet.Samples/IntroFirstCallColumn.cs)]`

AndreyAkinshin commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,74 @@
+        private static bool HasAnyRealWorkloads(BenchmarkReport report)
+            => report.AllMeasurements.Any(m => m.IterationMode == IterationMode.Workload);
+
+        private static Measurement GetFirstCall(BenchmarkReport report)
```
If `report.AllMeasurements` doesn't contain Workload measurements (may be happened because of a mistake in another place), `GetValue` will throw an exception. It would be better to handle such a case, and print a placeholder like `NA` or `?`.

adamsitnik replied:
the method is always used in a context where `HasAnyRealWorkloads` was executed first so it won't happen

AndreyAkinshin commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public UnitType UnitType => UnitType.Time;
+
+        public string Legend => "Execution time of the first call (Jitting included)";
```
I suggest removing the "(Jitting included)" note. It doesn't provide any value for people who are not aware of jitting and may confuse people who know about it (how is it possible to measure the duration of the first call without jitting?). Also, it will be better to write "first method call" or "first benchmark call" instead of just "first call".

adamsitnik replied:
Good point, without profiler, it's impossible to measure the duration of the first call without jitting. I've removed the text.

adamsitnik commented on src/BenchmarkDotNet/Columns/FirstCallColumn.cs:
```diff
@@ -0,0 +1,74 @@
+
+        public UnitType UnitType => UnitType.Time;
+
+        public string Legend => "Execution time of the first call (Jitting included)";
```
```suggestion
        public string Legend => "Execution time of the first call";
```
